sim/common/shared/shared_utils.go

```diff
@@ -382,6 +413,100 @@ func NewStackingStatBonusEffect(config StackingStatBonusEffect) {
 	})
 }
 
+type EnchantProcStatBonusEffect struct {
+	EnchantID int32
+	Slots     []proto.ItemSlot
+	ProcStatBonusEffect
+}
+
+func factory_EnchantStatBonusEffect(config EnchantProcStatBonusEffect, extraSpell func(agent core.Agent) ExtraSpellInfo) {
+	core.NewEnchantEffect(config.EnchantID, func(agent core.Agent) {
+		character := agent.GetCharacter()
+
+		procID := core.ActionID{SpellID: config.AuraID}
+		if procID.IsEmptyAction() {
+			procID = core.ActionID{ItemID: config.ID}
+		}
+		procAura := character.NewTemporaryStatsAura(config.Name+" Proc", procID, config.Bonus, config.Duration)
+
+		if config.PPM != 0 && config.ProcMask == core.ProcMaskUnknown {
+			config.ProcMask = character.GetProcMaskForEnchant(config.EnchantID)
+		}
+
+		var procSpell ExtraSpellInfo
+		if extraSpell != nil {
+			procSpell = extraSpell(agent)
+		}
+
+		handler := func(sim *core.Simulation, spell *core.Spell, result *core.SpellResult) {
+			procAura.Activate(sim)
+			if procSpell.Spell != nil {
+				procSpell.Trigger(sim, spell, result)
+			}
+		}
+
+		triggerAura := core.MakeProcTriggerAura(&character.Unit, core.ProcTrigger{
+			ActionID:   core.ActionID{SpellID: config.ID},
+			Name:       config.Name,
+			Callback:   config.Callback,
+			ProcMask:   config.ProcMask,
+			Outcome:    config.Outcome,
+			Harmful:    config.Harmful,
+			ProcChance: config.ProcChance,
+			PPM:        config.PPM,
+			ICD:        config.ICD,
+			Handler:    handler,
+		})
+
+		if config.ICD != 0 {
+			procAura.Icd = triggerAura.Icd
+		}
+
+		if config.PPM != 0 {
+			character.ItemSwap.RegisterPPMEffect(config.EnchantID, config.PPM, triggerAura.Ppmm, triggerAura)
+		} else {
+			character.ItemSwap.RegisterEnchantProc(config.EnchantID, triggerAura, config.Slots)
+		}
```

Comment:
Ok I think I see the problem here: RegisterPPMEffect() is intended for procs that will always be equipped at all times, but where the proc rate can change dynamically when the weapon slots are swapped (due to PPM normalization). As a result, this code will break if, for example, you weapon swap between a weapon that has Windwalk on it and another weapon with a different enchant. The code as currently written will allow Windwalk to still proc off the new weapon, just with an adjusted proc rate based on the new weapon's speed.

If I'm understanding the code correctly, the fix would be to always call RegisterEnchantProc() for both ppm and fixed proc rate effects, so that the aura will properly be deactivated if it's not present in the swap set, but then *additionally* call RegisterPPMEffect() when relevant so that proc rates are dynamically adjusted.